File: python/topKElements.py
from typing import List


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        length = len(nums)
        freq = [ [] for i in range(length+1)]
        counts = {}

        for n in nums:
            counts[n] = 1 + counts.get(n, 0)

        for key, val in counts.items():
            freq[val].append(key)

        res = []
        for i in range(length, 0, -1):
            for v in freq[i]:
                res.append(v)
                if len(res) == k:
                    return res
                    

File: TestLeetCode/Array/TopKElements.cs
/**************************************************************
https://leetcode.com/problems/top-k-frequent-elements/description/
https://www.youtube.com/watch?v=YPTqKIgVk-k

hint: bucket
**************************************************************/

public class TopKFrequentSolution
{
    public IList<int> TopKFrequent(int[] nums, int k)
    {
        int length = nums.Length;
        List<int>[] freq = new List<int>[length + 1];
        for (int i = 0; i < freq.Length; i++)
        {
            freq[i] = new List<int>();
        }

        Dictionary<int, int> counts = new Dictionary<int, int>();
        foreach (int num in nums)
        {
            if (counts.ContainsKey(num))
            {
                counts[num]++;
            }
            else
            {
                counts[num] = 1;
            }
        }

        foreach (var entry in counts)
        {
            freq[entry.Value].Add(entry.Key);
        }

        List<int> res = new List<int>();
        // Count is natually sorted already, so going from 
        // highest count to lowest
        for (int i = length; i > 0; i--)
        {
            foreach (int num in freq[i])
            {
                res.Add(num);
                if (res.Count == k)
                {
                    return res;
                }
            }
        }

        return res;
    }
}

